Error when required members are not set

333fred opened the pull request:
Implements reading the required members list of a type, and enforces that required members are all set by an initializer on the constructor. Required members must be initialized with values, not with nested object initializers.

Test plan https://github.com/dotnet/roslyn/issues/57046.
Specification https://github.com/dotnet/csharplang/blob/main/proposals/required-members.md

333fred commented:
@jcouv @RikkiGibson please take a look.

333fred commented:
As a general note, I scoured the codebase for other places where `new BoundObjectCreation` occurs, and everything else is invocations of the `Nullable` constructor. I don't believe we care about `Nullable<T>` being defined with required members, so I didn't bother to insert checks elsewhere for it.

Youssef1313 commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -5427,7 +5499,9 @@ private bool IsConstructorAccessible(MethodSymbol constructor, ref CompoundUseSi
                     hasError);
 
-                return result;
+                CheckRequiredMembersInObjectInitializer(creation, diagnostics);
```
Something similar needed for attributes?

```csharp
[My] // should be error.
[My(P = 0)] // should be okay.
public class MyAttribute : System.Attribute
{
    public required int P { get; set; }
}
```

RikkiGibson commented on src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:
```diff
@@ -1222,4 +1222,1388 @@ struct S3
+}
+";
+
+        var creationWithoutAssignment = @"
```
variable name looks inaccurate here.

RikkiGibson commented on src/Compilers/CSharp/Portable/Symbols/Source/SourceMemberContainerSymbol.cs:
```diff
@@ -2400,6 +2400,20 @@ private void CheckForRequiredMemberAttribute(BindingDiagnosticBag diagnostics)
+                    }
+
+                    // The required members list for the base type '{0}' is malformed and cannot be interpreted. To use this constructor, apply the `SetsRequiredMembers` attribute.
+                    diagnostics.Add(ErrorCode.ERR_RequiredMembersBaseTypeInvalid, method.Locations[0], BaseTypeNoUseSiteDiagnostics);
```
Should we 'break' after this?

333fred replied:
No. I'm intentionally reporting on every constructor.

RikkiGibson commented on src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:
```diff
@@ -1222,4 +1222,1388 @@ struct S3
+        // }
+
+        var ilSource = @"
+.assembly extern original {}
```
Do we need to use a record to meaningfully test the scenario here? Seems like IL would be much smaller without the record members.

333fred replied:
Yes. The record is the point of this test.

Youssef1313 commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -7007,6 +7007,18 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+    <value>The required members list for '{0}' is malformed and cannot be interpreted.</value>
+  </data>
+  <data name="ERR_RequiredMembersBaseTypeInvalid" xml:space="preserve">
+    <value>The required members list for the base type '{0}' is malformed and cannot be interpreted. To use this constructor, apply the `SetsRequiredMembers` attribute.</value>
```
> apply the `SetsRequiredMembers` attribute

Most diagnostic messages use `'`, not `` ` ``

333fred commented:
@RikkiGibson addressed your feedback. Please take another look.

RikkiGibson commented on src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:
```diff
@@ -2604,6 +2573,205 @@ class C : I
+        var code = @"
+using System;
+
+[Attr]
```
Consider updating the feature proposal to clearly identify this scenario as being supported by the feature.

333fred commented:
@jcouv for a second review.

jcouv commented on src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:
```diff
@@ -1222,4 +1223,1555 @@ struct S3
+    public void EnforcedRequiredMembers_NoInheritance_DisallowedNestedObjectInitializer()
+    {
+        var c = @"
+var c = new C() { D1 = { NestedProp = 1 }, D2 = { NestedProp = 2 } };
```
nit: consider adding the version that works for contrast: `var c = new C() { D1 = new D() { NestedProp = 1 }, D2 = new D() { NestedProp = 2 } };`

jcouv commented on src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:
```diff
@@ -1222,4 +1223,1555 @@ struct S3
+Imports System.Runtime.CompilerServices
+<RequiredMember>
+Public Class Base
+    <RequiredMember>
```
Are we going to disallow usage of this attribute in VB?

333fred replied:
I wasn't planning on it.

jcouv replied:
Added a note to test plan

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -5747,7 +5842,23 @@ private BoundExpression BindTypeParameterCreationExpression(ObjectCreationExpres
             }
 
-            diagnostics.Add(errorLocation, useSiteInfo);
+            if (suppressUnsupportedRequiredMembersError && useSiteInfo.AccumulatesDiagnostics && useSiteInfo.Diagnostics is { Count: not 0 })
```
I didn't understand the benefit of this suppression. If a constructor gets picked and it has a problem, why not report it?

333fred replied:
Because then you'll get the error, even if you decorate the constructor with `SetsRequiredMembers` (which will be supported in a later pr.

jcouv replied:
Thanks. Consider leaving a comment to that effect

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -4967,6 +4968,96 @@ private static void ReportDuplicateObjectMemberInitializers(BoundExpression boun
+
+            var requiredMembers = constructor.ContainingType.AllRequiredMembers.ToBuilder();
+
+            if (requiredMembers.Count == 0)
```
nit: Consider moving count check before `ToBuilder()` allocation

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -4967,6 +4968,96 @@ private static void ReportDuplicateObjectMemberInitializers(BoundExpression boun
+                    ObjectCreationExpressionSyntax { Type: { } type } => type.Location,
+                    BaseObjectCreationExpressionSyntax { NewKeyword: { } newKeyword } => newKeyword.GetLocation(),
+                    AttributeSyntax { Name: { } name } => name.Location,
+                    _ => creationSyntax.Location
```
Can we hit this arm?

jcouv commented on src/Compilers/CSharp/Portable/Binder/Binder_Expressions.cs:
```diff
@@ -4967,6 +4968,96 @@ private static void ReportDuplicateObjectMemberInitializers(BoundExpression boun
+                        BoundFieldAccess fieldAccess => fieldAccess.FieldSymbol,
+                        _ => null
+                    }, assignmentOperator.Right)
+                    : default;
```
nit: expression is hard to parse. Consider just grabbing the memberSymbol here, then grabbing the right part later on when we need it.

jcouv commented on src/Compilers/CSharp/Portable/CSharpResources.resx:
```diff
@@ -7007,6 +7007,18 @@ To remove the warning, you can use /reference instead (set the Embed Interop Typ
+  <data name="ERR_RequiredMemberMustBeSet" xml:space="preserve">
+    <value>Required member '{0}' must be set in the object initializer or attribute constructor.</value>
+  </data>
+  <data name="ERR_RequiredMembersMustBeAssignment" xml:space="preserve">
```
>MustBeAssignment [](http://example.com/codeflow?start=33&length=16)

nit: "MustBeAssignment" sounds off.

jcouv reviewed (commented):
Done with review pass (iteration 2). Will look at tests this afternoon

333fred commented on src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:
```diff
@@ -499,6 +509,118 @@ public override string MetadataName
+            else
+            {
+                // Ensure that, after this method, _lazyRequiredMembersErrorDoNotAccessDirectly is able to be accessed safely.
+                Interlocked.MemoryBarrier();
```
`_lazyRequiredMembers` is a struct. I can't really make a "sentinel" value with it, and thus can't use it to represent what I need.

jcouv replied:
Thanks. I was looking at the wrong `ImmutableSegmentedDictionary` (non-generic one is a class).
Check to see what Aleksey thinks, but I would rather wrap the struct and stay in a less error-prone concurrency pattern.

AlekseyTs replied:
Let's discuss this offline. I think we shouldn't invent new ways of handling lazy initialization. For structs we usually declare wrapper classes in order to ensure atomic read/write operations. See, for example, ```TypeWithAnnotations.Boxed```. I probably could be convinced otherwise if this situation is unique in some way.

333fred replied:
I'm ok with doing a `Boxed` struct or similar here. I was generally trying to avoid an allocation, but I'm fine with saving that for if we decide there's a perf issue.

jcouv commented on src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:
```diff
@@ -1222,4 +1223,1592 @@ struct S3
+
+public class C
+{
+    public required int Prop1 { get; }
```
Do we have a test with required unsettable property from metadata?

jcouv replied:
Gentle ping

333fred replied:
I added one, just below this one.

jcouv commented on src/Compilers/CSharp/Test/Symbol/Symbols/RequiredMembersTests.cs:
```diff
@@ -1222,4 +1223,1592 @@ struct S3
+        CompileAndVerify(originalComp, verify: ExecutionConditionUtil.IsCoreClr ? Verification.Passes : Verification.Skipped).VerifyDiagnostics();
+
+        // This IL is the equivalent of:
+        // public record Derived : Base
```
nit: For these `CompileIL`, I don't see the point of bloating the IL by making it a record. A class would have sufficed to the purpose of those tests.

333fred replied:
It would not. The point of these tests is for copy constructors and similar, and in a future PR I'm working on these being records pointed out some considerations I'll need to think about for how copy constructors work with `SetsRequiredMembers`. I'll keep these as is.

333fred commented:
@jcouv addressed feedback

@sharwell please review https://github.com/dotnet/roslyn/pull/60101/commits/31157e3121ff0b6178382317f8943a9c061e0fc8 for changes to `ImmutableSegmentedDictionary`.

jcouv reviewed (approved):
LGTM Thanks (iteration 6) with one test to consider

jcouv commented on src/Compilers/CSharp/Portable/Symbols/NamedTypeSymbol.cs:
```diff
@@ -499,6 +509,119 @@ public override string MetadataName
+        private bool TryCalculateRequiredMembers(ref ImmutableSegmentedDictionary<string, Symbol>.Builder? requiredMembersBuilder)
+        {
+            var lazyRequiredMembers = _lazyRequiredMembers;
+            if (!ImmutableSegmentedDictionary.ReferenceEquals(RequiredMembersSentinel, lazyRequiredMembers))
```
Cool solution. Much easier to follow now. Thanks!

sharwell commented on src/Dependencies/Collections/ImmutableSegmentedDictionary`2.cs:
```diff
@@ -70,7 +70,18 @@ namespace Microsoft.CodeAnalysis.Collections
+        /// </summary>
+        internal static ImmutableSegmentedDictionary<TKey, TValue> CreateEmpty() => new(new SegmentedDictionary<TKey, TValue>());
+
+        /// <summary>
+        /// Compares the given segmented dictionaries for reference-equality of their underlying storage.
+        /// </summary>
+        internal static bool ReferenceEquals(ImmutableSegmentedDictionary<TKey, TValue> left, ImmutableSegmentedDictionary<TKey, TValue> right) => ReferenceEquals(left._dictionary, right._dictionary);
```
This is already implemented as `operator==`.

sharwell commented on src/Dependencies/Collections/ImmutableSegmentedDictionary`2.cs:
```diff
@@ -70,7 +70,18 @@ namespace Microsoft.CodeAnalysis.Collections
-        public static readonly ImmutableSegmentedDictionary<TKey, TValue> Empty = new(new SegmentedDictionary<TKey, TValue>());
+        public static readonly ImmutableSegmentedDictionary<TKey, TValue> Empty = CreateEmpty();
+
+        /// <summary>
+        /// Creates a new empty segmented dictionary, that is not underlying-reference equal to <see cref="Empty"/>. The intent is for this to
+        /// be used for sentinel values only.
+        /// </summary>
+        internal static ImmutableSegmentedDictionary<TKey, TValue> CreateEmpty() => new(new SegmentedDictionary<TKey, TValue>());
```
📝 This doesn't seem like a generally-valuable API. For the edge case where it's needed, it would be better to construct the sentinel instance using private reflection rather than facilitate this type of usage in other scenarios.

jcouv reviewed (approved):
LGTM Thanks (iteration 7)
